### DISCUSSION TITLE
How to use question marks in tonic-build async method ? anyhow way or Result<Xxx, Box<dyn Error>>

### DISCUSSION THREAD

**Discussion by xiuno** [hyperium/tonic#716] [Q&A] [answered]
anyhow way, I can use question mark: 
```
fn fn1() -> Result<Xxx> {
        let a = "123".parse::<u32>()?;
        let b = "456".parse::<u32>()?;
        ...
}
```

but in tonic,  I have to map_err() any where...
```
#[tonic::async_trait]
impl Xxx for XxxXxx {
    async fn xxx(&self, request: Request<Empty>) -> Result<Response<Xxx>, Status> {
         let a = "123".parse::<u32>().map_err(|e|Status::new(Code::Xxxx, "xxx"))?;
         let b = "123".parse::<u32>().map_err(|e|Status::new(Code::Xxxx, "xxx"))?;
         ...
    }
}
```
is there better way ?

if I define MyError and change Status to MyError, will get error Status not match MyError in trait.

**Comment by davidpdrsn** [comment, marked as answer]
If the error type is local to your crate you can define `impl From<MyError> for Status` and then `?` will work directly.

Doesn't work for anyhow though. That requires using `map_err`. An `anyhow::Error` can happen for whatever reason, so tonic couldn't fine a general mapping. Therefore you're required to pick which status is appropriate with `map_err`.

**Reply by xiuno** [reply]
I tried that, but Result<Response<Xxx>, Status> and Result<Response<Xxx>, MyError> don't match ! Status is defined in trait, to modify it, may need to modify a lots of codes of tonic build.

**Reply by davidpdrsn** [reply]
This works:

```rust
#[tonic::async_trait]
impl Greeter for MyGreeter {
    async fn say_hello(
        &self,
        request: Request<HelloRequest>,
    ) -> Result<Response<HelloReply>, Status> {
        // try_thing returns Result<_, MyError> but can still use ?
        try_thing()?;

        // ...
    }
}

struct MyError {}

impl From<MyError> for Status {
    fn from(err: MyError) -> Status {
        // ...
    }
}

fn try_thing() -> Result<(), MyError> {
    // ...
}
```

**Reply by xiuno** [reply]
In this way, it can really work, wrapping a layer, and it would be better if it could be defined as a macro.

**Reply by davidpdrsn** [reply]
I don't understand what you mean. Can you clarify what you mean by

> [..] wrapping a layer, and it would be better if it could be defined as a macro.

**Reply by gcxfd** [reply]
@davidpdrsn @xiuno i write a macro https://crates.io/crates/tonic_catch , now can use like this 

```
use tonic_catch::{tonic_catch, Result, Error};

#[tonic_catch]
impl Rmw for RmwSrv {
  async fn head(&self, request: Request<Url>) -> Result<()> {
    let addr = match request.remote_addr() {
      Some(addr) => addr.to_string(),
      None => String::new(),
    };

    let url = request.into_inner();
    println!("{:?} {}/{}", addr, url.addr, url.path);

    self
      .sender
      .send(Evt::Head(url.addr.parse()?, url.path))
      .await?;

    Ok(Response::new(()))
  }
}
```

**Comment by gcxfd** [comment]
use this https://crates.io/crates/tonic_catch , everything will be easy

As shown in the figure below, I found that the answer was folded, and I paste it again for convenience

<img width="912" alt="image" src="https://user-images.githubusercontent.com/69608605/153552388-9c75f4f5-88ba-48b4-af2e-579d12075719.png">
